File: 11_动态规划/dp分类/状压dp/visited上携带参数/template.go
package main

func main() {
	var nums []int
	const INF int = 1e18
	n := len(nums)
	mask := 1<<n - 1

	memo := make([]int, 1<<n)
	for i := range memo {
		memo[i] = -1
	}
	hash := func(visited int) int {
		return visited
	}

	var dfs func(visited int) int
	dfs = func(visited int) int {
		if visited == mask {
			return 0
		}
		hash_ := hash(visited)
		if memo[hash_] != -1 {
			return memo[hash_]
		}
		res := INF
		for next := 0; next < n; next++ {
			if visited&(1<<next) == 0 {
				res = min(res, dfs(visited|(1<<next))+1)
			}
		}
		memo[hash_] = res
		return res
	}

}

File: 11_动态规划/dp分类/状压dp/visited上携带参数/template.py
from functools import lru_cache
from typing import List


def solve(nums: List[int]) -> int:
    n = len(nums)
    mask = (1 << n) - 1

    @lru_cache(None)
    def dfs(visited: int) -> int:
        if visited == mask:
            return 0
        res = INF
        for i in range(n):
            if (visited >> i) & 1:
                continue
            res = min2(res, dfs(visited | (1 << i)) + nums[i])

        return res

    res = dfs(0)
    dfs.cache_clear()
    return res
